common/src/main/java/tachyon/MasterClientBase.java

```diff
@@ -0,0 +1,190 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package tachyon;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.net.InetSocketAddress;
+import java.util.concurrent.ExecutorService;
+
+import org.apache.thrift.protocol.TBinaryProtocol;
+import org.apache.thrift.protocol.TMultiplexedProtocol;
+import org.apache.thrift.protocol.TProtocol;
+import org.apache.thrift.transport.TFramedTransport;
+import org.apache.thrift.transport.TSocket;
+import org.apache.thrift.transport.TTransportException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.base.Throwables;
+
+import tachyon.conf.TachyonConf;
+import tachyon.retry.ExponentialBackoffRetry;
+import tachyon.retry.RetryPolicy;
+import tachyon.util.network.NetworkAddressUtils;
+
+/**
+ * The base class for master clients.
+ */
+public abstract class MasterClientBase implements Closeable {
+  private static final Logger LOG = LoggerFactory.getLogger(Constants.LOGGER_TYPE);
+
+  protected final boolean mUseZookeeper;
+  protected final ExecutorService mExecutorService;
+  protected final TachyonConf mTachyonConf;
+
+  protected InetSocketAddress mMasterAddress = null;
+  protected TProtocol mProtocol = null;
+  protected volatile boolean mConnected;
+  protected volatile boolean mIsClosed;
+
+  public MasterClientBase(InetSocketAddress masterAddress, ExecutorService executorService,
+                          TachyonConf tachyonConf) {
+    mTachyonConf = tachyonConf;
+    mUseZookeeper = mTachyonConf.getBoolean(Constants.USE_ZOOKEEPER);
+    if (!mUseZookeeper) {
+      mMasterAddress = masterAddress;
+    }
+    mConnected = false;
+    mIsClosed = false;
+    mExecutorService = executorService;
+  }
+
+  /**
+   * Returns the name of the service.
+   *
+   * @return A string representing the specific master
+   */
+  protected abstract String getServiceName();
+
+  /**
+   * This called after the connection is made to the master. Here, implementations should create
+   * internal state to finish the connection process.
+   */
+  protected abstract void afterConnect();
+
+  /**
+   * This called after the connection is disconnected. Here, implementations should clean up any
+   * additional state created for the connection.
+   */
+  protected abstract void afterDisconnect();
+
+  /**
+   * Connect with the master; an exception is thrown if this fails.
+   *
+   * @throws IOException
+   */
+  public synchronized void connect() throws IOException {
+    if (mConnected) {
+      return;
+    }
+    disconnect();
+
+    if (mIsClosed) {
+      throw new IOException("Client is closed, will not try to connect");
+    }
+
+    Exception lastException = null;
```

Comment:
I don't think this is necessary, we are throwing up an IOException anyway (so no chance of the higher level catching a specific exception) and all the exceptions get logged.